File: IT/Programmazione funzionale/Esercizi/Python/esercizio14.py
# file esercizio14.py

from __future__ import print_function
from sys import version_info

if version_info[0] > 2: from functools import reduce

li = ["Nel", "bel", "mezzo", "del", "cammin", "di", "nostra", "vita"]

s = reduce(lambda t1, t2 : "{} {}".format(t1, t2), li)

print(li)
print("\"{}\"".format(s))

File: IT/Programmazione funzionale/Esercizi/C++/esercizio14.cpp
/*
file esercizio14.cpp
compilare con: g++ -std=c++11 esercizio14.cpp
*/

#include <iostream>
#include <iterator>
#include <algorithm>
#include <vector>
#include <numeric>

using std::cout;
using std::endl;
using std::ostream_iterator;
using std::string;
using std::vector;

int main()
{
	vector<string> v = {"Nel", "bel", "mezzo", "del", "cammin", "di", "nostra", "vita"};

	string s = accumulate(v.begin(), v.end(), string(), [](string& result, const string& t) {
		return !result.empty() ? result + " " + t : t;
	});

	cout << "[";
	copy(v.begin(), v.end(), ostream_iterator<string>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "\"" << s << "\"" << endl;
}

